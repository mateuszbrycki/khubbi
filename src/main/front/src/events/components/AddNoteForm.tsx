import React, {useState} from "react";
import {Col, Row} from "react-bootstrap";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import {EventDate} from "../../types";

export const DATE_FORMAT = "yyyy-MM-dd";
export const TIME_FORMAT = "HH:mm";

export interface AddNoteFormProps {
}

export interface AddNoteFormActionProps {
    readonly addNote: (note: string, date: EventDate) => void
    readonly currentDateProvider: () => EventDate
}

const AddNoteForm: React.FC<AddNoteFormProps & AddNoteFormActionProps> = (props) => {

    const {addNote, currentDateProvider} = props

    const [note, setNote] = useState<string>("")
    const [date, setDate] = useState<string>(currentDateProvider().format(DATE_FORMAT))
    const [time, setTime] = useState<string>(currentDateProvider().format(TIME_FORMAT))

    const resetFormValues = () => {
        setNote("");
        setDate(currentDateProvider().format(DATE_FORMAT));
        setTime(currentDateProvider().format(TIME_FORMAT));
    }

    return <Form onSubmit={event => {
        addNote(note, EventDate.ofDateAndTime(`${date}T${time}`));
        event.preventDefault();
        resetFormValues();
    }}>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="note">
                <Form.Control as="textarea"
                              data-testid="note-textarea"
                              rows={5}
                              value={note}
                              onChange={(event) => setNote(event.target.value)}
                />
            </Form.Group>
        </Row>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="noteAddDate">
                <Form.Control type="date"
                              data-testid="note-add-date"
                              value={date}
                              onChange={(event) => setDate(event.target.value)}
                />

            </Form.Group>
            <Form.Group as={Col} controlId="noteAddTime">
                <Form.Control type="time"
                              data-testid="note-add-time"
                              value={time}
                              onChange={(event) => setTime(event.target.value)}
                />
            </Form.Group>
            <Form.Group as={Col} controlId="submitForm">
                <Button type="submit" data-testid="add-note-button">Add note</Button>
            </Form.Group>
        </Row>
    </Form>

}

export default AddNoteForm
