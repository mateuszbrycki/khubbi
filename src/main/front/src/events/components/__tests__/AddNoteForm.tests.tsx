import AddNoteForm from "../AddNoteForm";
import {act, fireEvent, render} from "@testing-library/react";
import ShallowRenderer from "react-test-renderer/shallow";

const getTodayDate = () => new Date().toISOString().substr(0, 10);
const testCurrentDateProvider = () => "2024-07-23"

const addNoteMock = jest.fn(() => Promise.resolve());


it('renders form', () => {
    const renderer = ShallowRenderer.createRenderer()
    renderer.render(<AddNoteForm addNote={addNoteMock} currentDateProvider={testCurrentDateProvider}/>)

    expect(renderer.getRenderOutput()).toMatchSnapshot()
});

test('should call add note', async () => {
    const {getByTestId} = render(<AddNoteForm addNote={addNoteMock} currentDateProvider={testCurrentDateProvider}/>);

    await act(async () => {
        fireEvent.change(getByTestId("note-textarea"), {target: {value: "test-note"}})
        fireEvent.change(getByTestId("note-add-date"), {target: {value: "2023-03-12"}})

        getByTestId("add-note-button").click();
    })

    expect(addNoteMock).toHaveBeenCalledWith("test-note", "2023-03-12");
})

test('should setup today`s date', async () => {
    const {getByTestId} = render(<AddNoteForm addNote={addNoteMock} currentDateProvider={getTodayDate}/>);

    expect(getByTestId("note-add-date").getAttribute("value")).toEqual(getTodayDate());
})

test('should setup today`s date after submit', async () => {
    const {getByTestId} = render(<AddNoteForm addNote={addNoteMock} currentDateProvider={testCurrentDateProvider}/>);

    await act(async () => {
        fireEvent.change(getByTestId("note-textarea"), {target: {value: "test-note"}})
        fireEvent.change(getByTestId("note-add-date"), {target: {value: "2023-03-12"}})

        getByTestId("add-note-button").click();
    })

    expect(addNoteMock).toHaveBeenCalledWith("test-note", "2023-03-12");

    expect(getByTestId("note-add-date").getAttribute("value")).toEqual(testCurrentDateProvider());
})
