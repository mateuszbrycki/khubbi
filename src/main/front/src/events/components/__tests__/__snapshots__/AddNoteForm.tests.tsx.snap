// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form 1`] = `
<Form
  onSubmit={[Function]}
>
  <Row
    className="mb-3"
  >
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="note"
    >
      <FormControl
        as="textarea"
        data-testid="note-textarea"
        onChange={[Function]}
        rows={5}
        value=""
      />
    </FormGroup>
  </Row>
  <Row
    className="mb-3"
  >
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="noteAddDate"
    >
      <FormControl
        data-testid="note-add-date"
        onChange={[Function]}
        type="date"
        value="2024-07-23"
      />
    </FormGroup>
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="submitForm"
    >
      <Button
        data-testid="add-note-button"
        type="submit"
      >
        Add note
      </Button>
    </FormGroup>
  </Row>
</Form>
`;
